### PR TITLE
[Impeller] Wire Flutter's own `VulkanMemoryAllocator` implementation

### PR DESCRIPTION
**Pull request by iskakaushik**
fixes: https://github.com/flutter/flutter/issues/114030

### CODE REVIEW

File: ci/licenses_golden/licenses_flutter [hunk 1]
```diff
@@ -831,6 +831,10 @@ FILE: ../../../flutter/flow/surface_frame.cc
 FILE: ../../../flutter/flow/surface_frame.h
 FILE: ../../../flutter/flow/surface_frame_unittests.cc
 FILE: ../../../flutter/flow/texture_unittests.cc
+FILE: ../../../flutter/flutter_vma/flutter_skia_vma.cc
```
Comment (iskakaushik) [inline ci/licenses_golden/licenses_flutter:834]: licenses need to be fixed

File: flutter_vma/flutter_skia_vma.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+#include "flutter/flutter_vma/flutter_vma.h"
+
+#include "flutter/fml/memory/ref_ptr.h"
+#include "flutter/vulkan/procs/vulkan_proc_table.h"
+#include "third_party/skia/include/gpu/vk/GrVkBackendContext.h"
+
+namespace flutter {
+
+class FlutterSkiaVulkanMemoryAllocator : public skgpu::VulkanMemoryAllocator {
```
Comment (chinmaygarde) [inline flutter_vma/flutter_skia_vma.h:15]: Nit: Perhaps just call this `SkiaVulkanMemoryAllocator`?

File: flutter_vma/flutter_skia_vma.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      bool mustUseCoherentHostVisibleMemory);
+
+  fml::RefPtr<vulkan::VulkanProcTable> vk_proc_table_;
+  VmaAllocator allocator_;
+
+  // For host visible allocations do we require they are coherent or not. All
+  // devices are required to support a host visible and coherent memory type.
+  // This is used to work around bugs for devices that don't handle non coherent
+  // memory correctly.
+  bool must_use_coherent_host_visible_memory_;
```
Comment (chinmaygarde) [inline flutter_vma/flutter_skia_vma.h:67]: Disallow copy and assign here?

File: flutter_vma/flutter_skia_vma.cc [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+      info.preferredFlags = VK_MEMORY_PROPERTY_HOST_CACHED_BIT;
+      break;
+  }
+
+  if (must_use_coherent_host_visible_memory_ &&
+      (info.requiredFlags & VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT)) {
+    info.requiredFlags |= VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+  }
+  if (kDedicatedAllocation_AllocationPropertyFlag & allocationPropertyFlags) {
+    info.flags |= VMA_ALLOCATION_CREATE_DEDICATED_MEMORY_BIT;
```
Comment (chinmaygarde) [inline flutter_vma/flutter_skia_vma.cc:171]: This reminds me, perhaps we should do this for all images from the image decoder (the ui.Images)?

File: flutter_vma/flutter_skia_vma.cc [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+  }
+  if ((kLazyAllocation_AllocationPropertyFlag & allocationPropertyFlags) &&
+      BufferUsage::kGpuOnly == usage) {
+    info.preferredFlags |= VK_MEMORY_PROPERTY_LAZILY_ALLOCATED_BIT;
+  }
+
+  if (kPersistentlyMapped_AllocationPropertyFlag & allocationPropertyFlags) {
+    SkASSERT(BufferUsage::kGpuOnly != usage);
+    info.flags |= VMA_ALLOCATION_CREATE_MAPPED_BIT;
```
Comment (chinmaygarde) [inline flutter_vma/flutter_skia_vma.cc:180]: We should never needs this with Impeller.

File: flutter_vma/BUILD.gn [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import("//flutter/flutter_vma/config.gni")
+
+source_set("flutter_vma") {
+  sources = [
+    "flutter_vma.cc",
+    "flutter_vma.h",
+  ]
+
+  if (disable_vma_stl_shared_mutex) {
+    defines = [ "VMA_USE_STL_SHARED_MUTEX=0" ]
```
[the remark this answers is not in the document]
Comment (iskakaushik) [reply flutter_vma/BUILD.gn:14]: This means that they use non-stl mutexes. For context, Impeller doesn't use `flutter_skia_vma.cc` stuff at all. We just use `allocator_vk.cc` copy of the vma.

File: flutter_vma/flutter_skia_vma.cc (hunk 4 above)
Comment (iskakaushik) [reply flutter_vma/flutter_skia_vma.cc:171]: Noted, will fix this in `allocator_vk.cc`

File: flutter_vma/flutter_skia_vma.cc [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+  if (must_use_coherent_host_visible_memory_ &&
+      (info.requiredFlags & VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT)) {
+    info.requiredFlags |= VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+  }
+  if (kDedicatedAllocation_AllocationPropertyFlag & allocationPropertyFlags) {
+    info.flags |= VMA_ALLOCATION_CREATE_DEDICATED_MEMORY_BIT;
+  }
+  if ((kLazyAllocation_AllocationPropertyFlag & allocationPropertyFlags) &&
+      BufferUsage::kGpuOnly == usage) {
+    info.preferredFlags |= VK_MEMORY_PROPERTY_LAZILY_ALLOCATED_BIT;
```
[the remark this answers is not in the document]
Comment (iskakaushik) [reply flutter_vma/flutter_skia_vma.cc:175]: Noted, will fix this in `allocator_vk.cc`

**Comment by zanderso** [comment]
Is this flake a couple of commits past this change related? https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20Production%20Engine%20Drone/5722/overview

**Pull request merged**